src/workerd/jsg/async-context.h

```diff
@@ -0,0 +1,166 @@
+      kj::Maybe<StorageEntry> maybeStorageEntry = nullptr);
+  // Create a new AsyncContextFrame. If maybeParent is not specified, uses the current().
+  // If maybeStorageEntry is non-null, the associated storage cell in the new frame is
+  // set to the given value.
+
+  static v8::Local<v8::Function> wrap(Lock& js, v8::Local<v8::Function> fn,
+                                      kj::Maybe<AsyncContextFrame&> maybeFrame = nullptr,
+                                      kj::Maybe<v8::Local<v8::Value>> thisArg = nullptr);
+  // Associates the given JavaScript function with the given AsyncContextFrame, returning
+  // a wrapper function that will ensure appropriate propagation of the async context
+  // when the wrapper function is called. If maybeFrame is not specified, the current()
+  // frame is used.
+
+  static v8::Local<v8::Promise> wrap(Lock& js, v8::Local<v8::Promise> promise,
+                                     kj::Maybe<AsyncContextFrame&> maybeFrame = nullptr);
+  // Associates the given JavaScript promise with the given AsyncContextFrame, returning
+  // the same promise back. If maybeFrame is not specified, the current() frame is used.
+
+  static kj::Maybe<AsyncContextFrame&> tryUnwrap(Lock& js, V8Ref<v8::Promise>& promise);
+  static kj::Maybe<AsyncContextFrame&> tryUnwrap(Lock& js, v8::Local<v8::Promise> promise);
+  // Returns a reference to the AsyncContextFrame that was current when the JS Promise
+  // was created. When async context tracking is enabled, this should always return a
+  // non-null value.
+
+  struct Scope {
+    // AsyncContextFrame::Scope makes the given AsyncContextFrame the current in the
+    // stack until the scope is destroyed.
+    IsolateBase& isolate;
+    Scope(Lock& js, AsyncContextFrame& frame);
+    Scope(v8::Isolate* isolate, AsyncContextFrame& frame);
+    ~Scope() noexcept(false);
+    KJ_DISALLOW_COPY(Scope);
+  };
+
+  kj::Maybe<Value&> get(StorageKey& key);
+  // Retrieves the value that is associated with the given key.
+
+  struct StorageScope {
+    // Creates a new AsyncContextFrame with a new value for the given
+    // StorageKey and sets that frame as current for as long as the StorageScope
+    // is alive.
+    kj::Own<AsyncContextFrame> frame;
+    Scope scope;
+    // Note that the scope here holds a bare ref to the AsyncContextFrame so it
+    // is important that these member fields stay in the correct cleanup order.
+
+    StorageScope(Lock& js, StorageKey& key, Value store);
+    KJ_DISALLOW_COPY(StorageScope);
+  };
+
+private:
+  struct StorageEntryCallbacks {
+    StorageKey& keyForRow(StorageEntry& entry) const {
+      return *entry.key;
+    }
+
+    bool matches(const StorageEntry& entry, StorageKey& key) const {
+      return entry.key->hashCode() == key.hashCode();
+    }
+
+    uint hashCode(StorageKey& key) const {
+      return key.hashCode();
+    }
+  };
+
+  using Storage = kj::Table<StorageEntry, kj::HashIndex<StorageEntryCallbacks>>;
```

Comment:
This can just be a HashMap, if you give StorageKey the appropriate `hashCode()` and `operator==()` implementations.

Reply:
hmm... the table is storing `kj::Own<StorageKey>` and I'm using `StorageKey&` to find. `find(...)` is giving a compile error when I try using a HashMap here.

Reply:
Oh I see. That's annoying. One solution would be to make `StorageKey` be a value type that holds `Own<State>` or whatever internally. But I guess if you'd rather leave it as a `Table` I'm OK with that.

Reply:
Yeah let's just leave it as it for now. We can tweak it later if necessary.